src/txn/txn_rollback_to_stable.c

```diff
@@ -88,6 +103,80 @@ __rollback_abort_newer_insert(
               session, ins->upd, rollback_timestamp, &stable_update_found);
 }
 
+/*
+ * __col_insert_alloc --
+ *     Column-store insert: allocate a WT_INSERT structure and fill it in.
+ */
+static int
+__col_insert_alloc(
+  WT_SESSION_IMPL *session, uint64_t recno, u_int skipdepth, WT_INSERT **insp, size_t *ins_sizep)
+{
+    WT_INSERT *ins;
+    size_t ins_size;
+
+    /*
+     * Allocate the WT_INSERT structure and skiplist pointers, then copy the record number into
+     * place.
+     */
+    ins_size = sizeof(WT_INSERT) + skipdepth * sizeof(WT_INSERT *);
+    WT_RET(__wt_calloc(session, 1, ins_size, &ins));
+
+    WT_INSERT_RECNO(ins) = recno;
+
+    *insp = ins;
+    *ins_sizep = ins_size;
+    return (0);
+}
+
+/*
+ * __rollback_col_add_update --
+ *     Add the provided update to the head of the update list.
+ */
+static inline int
+__rollback_col_add_update(
+  WT_SESSION_IMPL *session, WT_PAGE *page, WT_COL *cip, WT_UPDATE *upd, uint64_t recno)
+{
+    WT_DECL_RET;
+    WT_INSERT *ins, **ins_stack[WT_SKIP_MAXDEPTH];
+    WT_INSERT_HEAD *ins_head, **ins_headp;
+    WT_PAGE_MODIFY *mod;
+    size_t ins_size, upd_size;
+    u_int skipdepth;
+
+    skipdepth = 1;
+
+    /* If we don't yet have a modify structure, we'll need one. */
+    WT_RET(__wt_page_modify_init(session, page));
+    mod = page->modify;
+
+    /* Allocate an update array as necessary. */
+    WT_PAGE_ALLOC_AND_SWAP(session, page, mod->mod_col_update, ins_headp, page->entries);
+
+    /* Set the WT_INSERT head array reference. */
+    ins_headp = &mod->mod_col_update[WT_COL_SLOT(page, cip)];
+    WT_PAGE_ALLOC_AND_SWAP(session, page, *ins_headp, ins_head, 1);
+    ins_head = *ins_headp;
+
+    /* Allocate an empty WT_INSERT for a recno. */
+    WT_ERR(__col_insert_alloc(session, recno, skipdepth, &ins, &ins_size));
+
+    /* Assign the update to the WT_INSERT. */
+    upd_size = __wt_update_list_memsize(upd);
+    ins->upd = upd;
+    ins_size += upd_size;
+
+    ins_stack[0] = &ins_head->head[0];
+    ins->next[0] = NULL;
+
+    /* Insert the new update at the head of the update list. */
+    WT_ERR(__wt_insert_serial(session, page, ins_head, ins_stack, &ins, ins_size, skipdepth, true));
+
+    return (ret);
```

Comment:
we should free the allocated `ins` in case of failure. Whereas the `upd` free is taken care of in the caller function.